src/controllers/simplified-account/settings/switch-psp/switch-to-worldpay/make-test-payment.controller.js

```diff
@@ -0,0 +1,74 @@
+const urljoin = require('url-join')
+const CREDENTIAL_STATE = require('@models/constants/credential-state')
+const worldpayDetailsService = require('@services/worldpay-details.service')
+const SELFSERVICE_URL = process.env.SELFSERVICE_URL
+
+function get (req, res) {
+  const account = req.account
+  const service = req.service
+  const context = {
+    backLink: formatSimplifiedAccountPathsFor(paths.simplifiedAccount.settings.switchPsp.switchToWorldpay.index,
+      service.externalId, account.type)
+  }
+  return response(req, res, 'simplified-account/settings/switch-psp/switch-to-worldpay/make-a-test-payment', context)
+}
+
+async function post (req, res, next) {
+  const account = req.account
+  const service = req.service
+  const targetCredential = account.getSwitchingCredential()
+  const chargeRequest = new ChargeRequest()
+    .withAmount(200)
+    .withDescription('Live payment to verify Worldpay integration')
+    .withReference('VERIFY_PSP_INTEGRATION')
+    .withReturnUrl(urljoin(SELFSERVICE_URL, formatSimplifiedAccountPathsFor(paths.simplifiedAccount.settings.switchPsp.switchToWorldpay.makeTestPayment.inbound,
+      service.externalId, account.type)))
+    .withCredentialId(targetCredential.externalId)
+    .withMoto(account.allowMoto)
+
+  chargeService.createCharge(service.externalId, account.type, chargeRequest)
+    .then((response) => {
+      req.session[VERIFY_PSP_INTEGRATION_CHARGE_EXTERNAL_ID_KEY] = response.charge_id
+      res.redirect(filterNextUrl(response))
+    })
+    .catch((error) => {
+      next(error)
+    })
+}
+
+async function getInbound (req, res, next) {
+  const account = req.account
+  const service = req.service
+  const user = req.user
+  const chargeExternalId = req.session[VERIFY_PSP_INTEGRATION_CHARGE_EXTERNAL_ID_KEY]
+  delete req.session[VERIFY_PSP_INTEGRATION_CHARGE_EXTERNAL_ID_KEY]
+  const targetCredential = account.getSwitchingCredential()
+  if (!chargeExternalId) {
+    throw new Error('No charge found on session')
+  }
+  chargeService.getCharge(service.externalId, account.type, chargeExternalId)
+    .then(async (charge) => {
+      if (charge.state.status === 'success') {
+        await worldpayDetailsService.updateCredentialState(service.externalId, account.type, targetCredential.externalId, user.externalId, CREDENTIAL_STATE.VERIFIED)
+        req.flash('messages', { state: 'success', icon: '&check;', heading: 'Payment verified', body: `This service is ready to switch to ${formatPSPName(targetCredential.paymentProvider)}` })
+      } else {
+        req.flash('messages', { state: 'error', heading: 'There is a problem', body: 'The payment has failed. Check your Worldpay credentials and try again. If you need help, contact govuk-pay-support@digital.cabinet-office.gov.uk' })
+      }
+      res.redirect(formatSimplifiedAccountPathsFor(paths.simplifiedAccount.settings.switchPsp.switchToWorldpay.index,
+        service.externalId, account.type))
+    })
+    .catch(err => next(err))
```

Comment:
this might be neater without the promise/then and just using async/await like

```javascript
const charge = await chargeService.getCharge(...)
if (charge.state.status !== 'success') {
 // return error
}
await worldpayDetailsService.updateCredentialState(...)
// return success
```
may need a try/catch to be sure errors are handled properly

Reply:
I'll experiment with the structure here once I have a cypress test for this controller